indexer/src/main/java/bio/terra/tanagra/indexing/job/ComputeAgeAtOccurrence.java

```diff
@@ -0,0 +1,409 @@
+
+import bio.terra.tanagra.indexing.BigQueryIndexingJob;
+import bio.terra.tanagra.indexing.job.beam.BigQueryUtils;
+import bio.terra.tanagra.query.*;
+import bio.terra.tanagra.underlay.AttributeMapping;
+import bio.terra.tanagra.underlay.Entity;
+import bio.terra.tanagra.underlay.Relationship;
+import bio.terra.tanagra.underlay.Underlay;
+import bio.terra.tanagra.underlay.Underlay.MappingType;
+import bio.terra.tanagra.utils.JavaUtils;
+import com.google.api.services.bigquery.model.TableRow;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.MoreCollectors;
+import com.google.common.collect.Streams;
+import java.time.Duration;
+import java.time.LocalDate;
+import java.time.Period;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Write.WriteDisposition;
+import org.apache.beam.sdk.transforms.FlatMapElements;
+import org.apache.beam.sdk.transforms.GroupByKey;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.join.CoGbkResult;
+import org.apache.beam.sdk.transforms.join.CoGroupByKey;
+import org.apache.beam.sdk.transforms.join.KeyedPCollectionTuple;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.beam.sdk.values.TypeDescriptors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Populate age_at_occurrence column. This column only exists for occurrence entities of
+ * CRITERIA_OCCURRENCE entity groups.
+ *
+ * <p>This job must run after DenormalizeEntityInstances, which writes the index occurrence table
+ * (minus age_at_occurrence column). This job reads in the entire table, populates age_at_occurrence
+ * column, and writes back table. (BQ doesn't let you modify existing rows, so we read and write
+ * entire table.)
+ */
+public class ComputeAgeAtOccurrence extends BigQueryIndexingJob {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ComputeAgeAtOccurrence.class);
+
+  private final Relationship occurrencePrimaryRelationship;
+
+  public ComputeAgeAtOccurrence(Entity entity, Relationship occurrencePrimaryRelationship) {
+    super(entity);
+    this.occurrencePrimaryRelationship = occurrencePrimaryRelationship;
+  }
+
+  @Override
+  public String getName() {
+    return "COMPUTE AGE AT OCCURRENCE (" + getEntity().getName() + ")";
+  }
+
+  @Override
+  public void run(boolean isDryRun) {
+    // Wait for DenormalizeEntityInstances to run.
+    if (!isDryRun) {
+      LOGGER.info("Waiting for DenormalizeEntityInstances to run for {}", getEntity().getName());
+      JavaUtils.retryUntilTrue(
```

Comment:
This is only 75s. If they're triggered at the same time, couldn't it take way longer than that for DenormalizeEntityInstances to complete? Or am I misunderstanding?

Reply:
Good point. Just now, I indexed `meaurement_occurrence`, which is one of the biggest tables with 1.4 billion rows and  17G (physical bytes). `ComputeAtAtOccurrence` failed because `DenormalizeEntityInstances` wasn't done in time. `DenormalizeEntityInstances` took 3-4 minutes. I changed this retry for 5 mins.